File: 3_Vigenere_cipher/Vigenere.cpp
/*
* Vigenère Cipher (Poly-alphabetic Cipher)
* Author: Vaasudevan (github.com/VaasuDevanS)
* Date: Dec 21, 2018
* Link: http://crypto.interactive-maths.com/vigenegravere-cipher.html
*/

#include <iostream>
#include <cmath>    // abs
using namespace std;

string encrypt(string input, string key)
{
    string encrypted = "";

    // Generate key_stream (match key_phrase with input_length)
    int input_len=input.size(), key_len=key.size();
    int diff = abs(input_len-key_len);
    string key_stream = "";
    for(int i=0; i<diff; i++)
        key_stream += key;
    key_stream = key_stream.substr(0, input_len);

    // Generate Tabula_Recta
    string Al = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"; Al += Al;
    char table[100][100];
    for(int i=65; i<65+26; i++) {
        string A_i = Al.substr(Al.find(i), 26);
        int k = 0;
        for(int j=65; j<65+27; j++) {
            table[i][j] = A_i[k++];
        }
    }

    /*
    // Print the Tabula_Recta
    cout << "   ";
    for(int i=65; i<65+26; i++)
        cout << (char)i << " ";
    cout << endl;
    for(int i=65; i<65+26; i++) {
        cout << (char)i << " ";
        for(int j=65; j<65+27; j++){
            cout << table[i][j] << " ";
        }
        cout <<endl;
    }
    */

    // Encrypt
    for(int i=0; i<input.size(); i++) {
        encrypted += table[input[i]][key_stream[i]];
    }

    return(encrypted);
}

string decrypt(string enc, string key)
{
    string decrypted = "";

    // Generate key_stream (match key_phrase with input_length)
    int enc_len=enc.size(), key_len=key.size();
    int diff = abs(enc_len-key_len);
    string key_stream = "";
    for(int i=0; i<diff; i++)
        key_stream += key;
    key_stream = key_stream.substr(0, enc_len);

    // Generate Tabula_Recta
    string Al = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"; Al += Al;
    char table[100][100];
    for(int i=65; i<65+26; i++) {
        string A_i = Al.substr(Al.find(i), 26);
        int k = 0;
        for(int j=65; j<65+26; j++) {
            table[j][A_i[k++]] = i;
        }
    }

    // DECRYPT
    for(int i=0; i<enc.size(); i++) {
        decrypted += table[key_stream[i]][enc[i]];
    }

    return(decrypted);
}

int main()
{
    string input, encrypted, decrypted, key="AWESOMECIPHER";
    cout << "Enter the String (only alpha): ";
    getline(cin, input);

    encrypted = encrypt(input, key);     // Call Encrypt
    decrypted = decrypt(encrypted, key); // Decrypt the Encrypted

    // Print the Encrypted and the Decrypted
    cout << "Encrypted: " << encrypted << endl;
    cout << "Decrypted: " << decrypted << endl;
}

// EOF

File: 3_Vigenere_cipher/Vigenere.py
"""
Vigenère Cipher (Poly-alphabetic Cipher)
"""
__author__ = "Vaasudevan (github.com/VaasuDevanS)"
__date__   = "Dec 21, 2018"
__link__   = "http://crypto.interactive-maths.com/vigenegravere-cipher.html"

from string import punctuation as P, digits as d
import re


def vigenere_encrypt(sent, key_phrase):

    # Keep-track for the spaces, non-alphanumeric and numeric chars
    extras = [(ix,c) for ix,c in enumerate(sent) if c in " "+P+d]
    
    # Generate key_stream
    sent = re.sub("[\W\d]", "", sent)
    sen_len, key_len = len(sent), len(key_phrase)
    key_stream = (key_phrase*abs(sen_len-key_len))[:sen_len]

    # Generate Tabula_Recta
    table = {}
    alphas = "".join((chr(i) for i in range(65, 65+26))) * 2
    alphabets = lambda c: alphas[alphas.index(c) : alphas.index(c)+26]
    for i in range(65, 65+26):
        for k,v in zip(range(65, 65+26), alphabets(chr(i))):
            table[chr(i), chr(k)] = v
    
    # Encrypt based on the table and insert the spaces
    encrypted = [table[(p,k)] for p,k in zip(sent, key_stream)]
    for ix, c in extras:
        encrypted.insert(ix, c)

    return "".join(encrypted)


def vigenere_decrypt(enc, key_phrase):

    # Keep-track for the spaces, non-alphanumeric and numeric chars
    extras = [(ix,c) for ix,c in enumerate(sent) if c in " "+P+d]

    # Generate key_stream
    enc = re.sub("[\W\d]", "", enc)
    enc_len, key_len = len(enc), len(key_phrase)
    key_stream = (key_phrase*abs(enc_len-key_len))[:enc_len]

    # Generate Tabula_Recta
    table = {}
    alphas = "".join((chr(i) for i in range(65, 65+26))) * 2
    alphabets = lambda c: alphas[alphas.index(c):alphas.index(c)+26]
    for i in range(65, 65+26):
        for k,v in zip(range(65, 65+26), alphabets(chr(i))):
            table[chr(k), v] = chr(i)

    # Decrypt based on the table
    decrypted = [table[(k,e)] for e,k in zip(enc, key_stream)]
    for ix, c in extras:
        decrypted.insert(ix, c)

    return "".join(decrypted)


if __name__ == "__main__":

    # Keyword for this Cipher
    key_phrase = "AwesomeCipher".upper()

    # Encrypt
    sent = input("Enter the String: ").upper()
    E = vigenere_encrypt(sent, key_phrase)

    # Decrypt
    D = vigenere_decrypt(E, key_phrase)

    print("Encrypted: %s\nDecrypted: %s" %(E, D))

# EOF
